datafusion/src/scalar.rs

```diff
@@ -831,13 +892,40 @@ impl ScalarValue {
                     "Unsupported creation of {:?} array from ScalarValue {:?}",
                     data_type,
                     scalars.peek()
-                )))
+                )));
             }
         };
 
         Ok(array)
     }
 
+    fn iter_to_decimal_array(
+        scalars: impl IntoIterator<Item = ScalarValue>,
+        precision: &usize,
+        scale: &usize,
+    ) -> Result<DecimalArray> {
+        // collect the value as Option<i128>
+        let array = scalars
+            .into_iter()
+            .map(|element: ScalarValue| match element {
+                ScalarValue::Decimal128(v1, _, _) => v1,
+                _ => unreachable!(),
+            })
+            .collect::<Vec<Option<i128>>>();
```

Comment:
Using `DecimalBuilder` in this way is awkward. I wonder if we could add a function to create Decimal values from an array of `i128` (likely it would go in arrow-rs). Something like

```
let arr = DecimalArray::from_iter_and_scale(array.into_iter(), precision, scale)
```

Reply:
> 

Yes, we can add the `from_iter_and_scale` function in arrow-rs later and replace this in the followup pull request.
Do you agree this? @alamb

Reply:
related pr: https://github.com/apache/arrow-rs/issues/1009